contracts/MetadataViews.cdc

```diff
@@ -0,0 +1,59 @@
+/**
+
+This contract implements the metadata standard proposed
+in FLIP-0636.
+
+Ref: https://github.com/onflow/flow/blob/master/flips/20210916-nft-metadata.md
+
+Structs and resources can implement one or more
+metadata types, called views. Each view type represents
+a different kind of metadata, such as a creator biography
+or a JPEG image file.
+*/
+
+pub contract MetadataViews {
+
+    // A ViewResolver provides access to a set of metadata views.
+    //
+    // A struct or resource (e.g. an NFT) can implement this interface
+    // to provide access to the views that it supports.
+    //
+    pub resource interface Resolver {
+        pub fun getViews(): [Type]
+        pub fun resolveView(_ view: Type): AnyStruct?
+    }
+
+    // Display is a basic view that includes the name and description
+    // of an object. Most objects should implement this view.
+    //
+    pub struct Display {
+        pub let name: String
+        pub let description: String
+
+        init(
+            name: String,
+            description: String,
+        ) {
+            self.name=name
+            self.description=description
+        }
+    }
+
+    // Thumbnail returns a thumbnail image for an object.
+    //
+    // Many NFT resources implement this view to provide 
+    // a simple visual representation of the NFT.
+    //
+    pub struct Thumbnail {
```

Comment:
Personally i do not like that this and Display is not split in two. Both for .find and blocto this would be that we have to do almost 2x the operations and grow complexity. I would rather the Display struct was even thicker. 

Fields that find needs to display sometehing "propperly" are
 - title
 - thumbnailUrl
 - sourceLink (in many cases this can be derrived from the id, so either we need id here or identity)
 - source (but we can get that from the collection)
 - tags (for filtereing)
 
 Also the Alchemy api has the following fields
 `pub let contract: NFTContract
    pub let id: UInt64
    pub let uuid: UInt64?
    pub let title: String?
    pub let description: String?
    pub let external_domain_view_url: String?
    pub let token_uri: String?
    pub let media: [NFTMedia?]
    pub let metadata: {String: AnyStruct}`

from https://github.com/alchemyplatform/alchemy-flow-contracts/blob/main/src/cadence/scripts/getNFTs.cdc

Reply:
Yeah, the `Thumbnail` struct is trading convenience for flexibility. I think we have two goals:

- Put all the basic info in one place (reduce complexity)
- Support multiple image formats and storage locations (flexibility)

IMO a simple string field is too limiting. For example, I think a web image (i.e. an HTTP URI) should be represented differently from an IPFS image (i.e. a CID). Some people might want to do on-chain SVGs. In the future, maybe some NFTs will use something entirely new.

It's trivial to implement all of these options as individual views. But as you pointed out, that pushes complexity onto the caller; they have to figure out which thumbnail views an NFT implements. 

I wondered if we could solve this with a `thumbnail` field on `Display` that points to a new interface.

```swift
pub struct Display {
   pub let thumbnail: Thumbnail
}

pub struct interface Thumbnail {
  // Problem: what if the image cannot (or should not) be represented with a URI?
  fun uri(): String
}
```

But I don't think that works. We're really just back to the string problem -- what common interface can represent all thumbnail types? There isn't one, really.

In the end I think the flexibility of multiple structs is worth the extra work needed to build a complete NFT representation. When you say 2x the operations, are you worried about transaction gas fees or just general code complexity? I was thinking it'd be fairly easy to add a helper function that resolves the most common thumbnail views and returns all the data as a single struct.

Reply:
I'm 👍 for `sourceLink` but I think `source` should always come from `nft.getType()` to avoid counterfeits.

Reply:
But the type of the nft does not have enough information to distinguish sources when using a shared minter. Take the example with counterfeit NFT socks from flowverse. If somebody else mints a sock on Rarible there is no way of distinguish them since they both used the same shared minter.

Reply:
> But I don't think that works. We're really just back to the string problem -- what common interface can represent all thumbnail types? There isn't one, really.

I do not see a better solution then a string.